src/lib/helpers/parse.js

```diff
@@ -0,0 +1,176 @@
+
+  constructor (src, privateKey = null, processEnv = process.env) {
+    this.src = src
+    this.privateKey = privateKey
+    this.processEnv = processEnv
+
+    //
+    this.parsed = {}
+    this.warnings = []
+
+    // for use with progressive expansion
+    this.runningParsed = {}
+  }
+
+  run () {
+    const lines = this.getLines()
+
+    let match
+    while ((match = Parse.LINE.exec(lines)) !== null) {
+      const key = match[1]
+      const value = match[2]
+      const quote = this.quote(value) // must be raw match
+      this.parsed[key] = this.clean(value, quote) // file value
+
+      // handle existing value - process.env wins
+      this.parsed[key] = this.processEnv[key] || this.parsed[key]
+
+      // decrypt
+      try {
+        this.parsed[key] = this.decrypt(this.parsed[key])
+      } catch (e) {
+        this.warnings.push(this.warning(e, key))
+      }
+
+      // eval empty, double, or backticks
+      if (quote !== "'" && (!this.inProcessEnv(key) || this.processEnv[key] === this.parsed[key])) {
+        this.parsed[key] = this.eval(this.parsed[key])
+      }
+
+      // expand empty, double, or backticks
+      if (quote !== "'") {
+        this.parsed[key] = resolveEscapeSequences(this.expand(this.parsed[key]))
+      }
+
+      // for use with progressive expansion
+      this.runningParsed[key] = this.parsed[key]
+    }
+
+    return {
+      parsed: this.parsed
+    }
+  }
+
+  trimmer (value) {
+    // Default undefined or null to empty string
+    return (value || '').trim()
+  }
+
+  quote (value) {
+    const v = this.trimmer(value)
+    const maybeQuote = v[0]
+    let q = ''
+    switch (maybeQuote) {
+      // single
+      case "'":
+        q = "'"
+        break
+      // double
+      case '"':
+        q = '"'
+        break
+      // backtick
+      case '`':
+        q = '`'
+        break
+      // empty
+      default:
+        q = ''
+    }
+
+    return q
+  }
+
+  clean (value, _quote) {
+    let v = this.trimmer(value)
+
+    // Remove surrounding quotes
+    v = v.replace(/^(['"`])([\s\S]*)\1$/mg, '$2')
+
+    // Expand newlines if double quoted
+    if (_quote === '"') {
+      v = v.replace(/\\n/g, '\n')
+      v = v.replace(/\\r/g, '\r')
+    }
+
+    return v
+  }
+
+  decrypt (value) {
+    return decryptValue(value, this.privateKey)
+  }
+
+  eval (value) {
+    const matches = value.match(/\$\([^()]+\)/) || []
+
+    return matches.reduce(function (newValue, match) {
+      // get command
+      const command = match.substring(2, match.length - 1)
+      // execute command
+      const value = chomp(execSync(command).toString())
+      // replace with command value
+      return newValue.replace(match, value)
+    }, value)
+  }
+
+  expand (value) {
+    const env = { ...this.runningParsed, ...this.processEnv } // current parsed and current process.env with process.env winning as default
```

Comment:
Shouldn't runningParsed win?

Reply:
i see your point but I think in this case the result should just be `foo` because it's already been set at process.env.

I believe that's also what docker-compose does.

Only in-file keys should be able to recursively update themselves I think.